ec/ecresource/deploymentresource/planmodifiers/topology_dedicated_masters.go

```diff
@@ -0,0 +1,189 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package planmodifiers
+
+import (
+	"context"
+	"github.com/elastic/cloud-sdk-go/pkg/api/deploymentapi/deploymentsize"
+	"github.com/elastic/cloud-sdk-go/pkg/models"
+	es "github.com/elastic/terraform-provider-ec/ec/ecresource/deploymentresource/elasticsearch/v2"
+	"github.com/elastic/terraform-provider-ec/ec/internal/util"
+	"github.com/hashicorp/terraform-plugin-framework/path"
+	"github.com/hashicorp/terraform-plugin-framework/resource"
+	"github.com/hashicorp/terraform-plugin-framework/types"
+	"github.com/hashicorp/terraform-plugin-framework/types/basetypes"
+	"github.com/hashicorp/terraform-plugin-log/tflog"
+)
+
+var objectAsOptions = basetypes.ObjectAsOptions{UnhandledNullAsEmpty: false, UnhandledUnknownAsEmpty: false}
+
+func UpdateDedicatedMasterTier(
+	ctx context.Context,
+	req resource.ModifyPlanRequest,
+	resp *resource.ModifyPlanResponse,
+	template models.DeploymentTemplateInfoV2,
+) {
+	var planElasticsearch es.ElasticsearchTF
+	resp.Diagnostics.Append(req.Plan.GetAttribute(ctx, path.Root("elasticsearch"), &planElasticsearch)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	dedicatedMastersThreshold := getDedicatedMastersThreshold(template)
+	if dedicatedMastersThreshold == 0 {
+		// No automatic dedicated masters management
+		return
+	}
+
+	nodesInCluster := countNodesInCluster(ctx, planElasticsearch)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	var config es.ElasticsearchTF
+	resp.Diagnostics.Append(req.Config.GetAttribute(ctx, path.Root("elasticsearch"), &config)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	if !config.MasterTier.IsNull() && !config.MasterTier.IsUnknown() {
```

Comment:
Should this check be moved to the top of this function? It feels like an opportunity to exit early before counting cluster nodes etc.